### PR TITLE
Modified scheduler according to updated design

### PR DESCRIPTION
**Pull request by kimeunju108**
What type of PR is this?

/kind feature

What this PR does / why we need it:
This PR presents the following updates of scheduler as follows:
- Push communication between resource collector and scheduler
- 2 Types of scheduling resources : Static & Dynamic
- Updated cluster & pod informers

Which issue(s) this PR fixes:

Fixes #

Special notes for your reviewer:

Does this PR introduce a user-facing change?:
NONE

### CODE REVIEW

File: globalscheduler/pkg/scheduler/common/apiserver/server.go [hunk 1]
```diff
@@ -41,23 +41,13 @@ type HTTPServer struct {
...
-
 // NewHTTPServer construct a new http server with ssl certificate
-func NewHTTPServer() (*HTTPServer, error) {
+func NewHTTPServer(ip string, port string) (*HTTPServer, error) {
 	hs := &HTTPServer{}
 
-	httpAddr := getServerAddress()
+	//httpAddr := getServerAddress()
+	httpAddr := fmt.Sprintf("%s:%s", "", port)
+	klog.Infof("http(https) listen: %s", httpAddr)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/common/apiserver/server.go:50]: Please replace port with http(https)

File: globalscheduler/pkg/scheduler/common/apiserver/server.go [hunk 2]
```diff
@@ -144,9 +136,11 @@ func RunServer(
 			return nil
 		}
 		fmt.Println("continue")
+		klog.Infof("***server continue: %v", msg)
 	default:
 		errMsg := fmt.Sprintf("%s due to error: %v", msg, err.Error())
 		fmt.Println(errMsg)
+		klog.Infof("***http server error: %v", errMsg)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/common/apiserver/server.go:143]: Log error with Errorf not Infof

File: globalscheduler/pkg/scheduler/eventhandlers.go [hunk 3]
```diff
@@ -237,6 +275,7 @@ func getStackSelector(selector *v1.ResourceSelector) types.Selector {
 
 func (sched *Scheduler) addPodToSchedulingQueue(obj interface{}) {
 	pod, ok := obj.(*v1.Pod)
+	klog.Infof("addPodToSchedulingQueue : %v", pod)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/eventhandlers.go:278]: Remove Infof

File: globalscheduler/pkg/scheduler/eventhandlers.go [hunk 4]
```diff
@@ -332,7 +373,9 @@ func (sched *Scheduler) skipStackUpdate(stack *types.Stack) bool {
 }
 
 func (sched *Scheduler) bindStacks(assumedStacks []types.Stack) {
+	klog.V(3).Infof("*** assumedStacks: %v", assumedStacks)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/eventhandlers.go:376]: V(3) does not work here

File: globalscheduler/pkg/scheduler/eventhandlers.go [hunk 5]
```diff
@@ -378,6 +421,7 @@ func (sched *Scheduler) bindToSite(clusterName string, assumedStack *types.Stack
 		},
 	}
 
+	klog.V(3).Infof("*** binding: %v", binding)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/eventhandlers.go:424]: V(3) does not work here

File: globalscheduler/pkg/scheduler/eventhandlers.go [hunk 6]
```diff
@@ -390,28 +434,98 @@ func (sched *Scheduler) bindToSite(clusterName string, assumedStack *types.Stack
...
+	sched.ClusterQueue.Add(KeyWithEventType{Key: key, EventType: eventType})
+}
+
+func (sched *Scheduler) verifyClusterInfo(cluster *clusterv1.Cluster) (verified bool) {
+	verified = false
+	ipAddress := cluster.Spec.IpAddress
+	region := cluster.Spec.Region.Region
+	az := cluster.Spec.Region.AvailabilityZone
+	clusterName := cluster.Name
+	if ipAddress == "" || region == "" || az == "" || clusterName == "" {
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/eventhandlers.go:525]: GeoLocation is not empty

File: globalscheduler/pkg/scheduler/eventhandlers.go [hunk 7]
```diff
@@ -390,28 +434,98 @@ func (sched *Scheduler) bindToSite(clusterName string, assumedStack *types.Stack
...
+		klog.Infof("cluster ipAddress:%s, region:%s, az:%s, or custer name:%s is null", ipAddress, region, az, clusterName)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/eventhandlers.go:526]: cluster not custer

File: globalscheduler/pkg/scheduler/factory/factory.go [hunk 8]
```diff
@@ -53,3 +53,16 @@ func NewPodInformer(schedulerName string, client clientset.Interface,
 			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
 	}
 }
+
+// NewPodInformer creates a shared index informer that returns only non-terminal pods.
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/factory/factory.go:57]: Removed it

File: globalscheduler/pkg/scheduler/framework/plugins/defaultbinder/default_binder.go [hunk 9]
```diff
@@ -66,7 +66,7 @@ func (b DefaultBinder) Bind(ctx context.Context, state *interfaces.CycleState, s
 		klog.Errorf("GetSiteSelectorState failed! err: %s", err)
 		return interfaces.NewStatus(interfaces.Error, fmt.Sprintf("getting site %q info failed: %v", siteID, err))
 	}
-
+	klog.Errorf("GetSiteSelectorState ===> %v", siteSelectedInfo)
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/framework/plugins/defaultbinder/default_binder.go:69]: Errof? Maybe Infof

File: globalscheduler/pkg/scheduler/framework/plugins/defaultbinder/default_binder.go [hunk 10]
```diff
@@ -100,3 +100,56 @@ func (b DefaultBinder) Bind(ctx context.Context, state *interfaces.CycleState, s
 
 	return nil
 }
+
+// Bind binds pods to site using the k8s client.
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/framework/plugins/defaultbinder/default_binder.go:104]: Remove it

File: globalscheduler/pkg/scheduler/framework/plugins/locationandoperator/locationandoperator.go [hunk 11]
```diff
@@ -46,27 +46,58 @@ func (pl *LocationAndOperator) locationEqual(stack *types.Stack, siteCacheInfo *
...
 	city := stack.Selector.GeoLocation.City
-	if city != "" && city != siteCacheInfo.GetSite().City {
+	if city != "" && city != siteCacheInfo.GetSite().GeoLocation.City {
 		return false, interfaces.NewStatus(interfaces.Unschedulable, "stack city not equal site city.")
 	}
 
 	return true, nil
 }
 
+/*func (pl *LocationAndOperator) locationEqual(stack *types.Stack, siteCacheInfo *sitecacheinfo.SiteCacheInfo) (bool, *interfaces.Status) {
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/framework/plugins/locationandoperator/locationandoperator.go:72]: Remove it

File: globalscheduler/pkg/scheduler/internal/cache/snapshot.go [hunk 12]
```diff
@@ -48,11 +47,14 @@ var _ schedulerlisters.SharedLister = &Snapshot{}
...
 		SiteCacheInfoMap: make(map[string]*schedulersitecacheinfo.SiteCacheInfo),
+		//SiteCacheInfoList: make([]*schedulersitecacheinfo.SiteCacheInfo, 0, len(SiteCacheInfoMap)),
+		RegionFlavorMap: make(map[string]*typed.RegionFlavor),
+		FlavorMap:       make(map[string]*typed.RegionFlavor),
 	}
 }
 
 // NewSnapshot initializes a Snapshot struct and returns it.
-func NewSnapshot(stacks []*types.Stack, sites []*types.Site) *Snapshot {
+/*func NewSnapshot(stacks []*types.Stack, sites []*types.Site) *Snapshot {
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/internal/cache/snapshot.go:57]: Remove it

File: globalscheduler/pkg/scheduler/internal/cache/snapshot.go [hunk 13]
```diff
@@ -105,9 +107,10 @@ func (s *Snapshot) SiteCacheInfos() schedulerlisters.SiteCacheInfoLister {
 }
 
 // NumSites returns the number of siteIDs in the snapshot.
-func (s *Snapshot) NumSites() int {
-	return len(s.SiteCacheInfoList)
-}
+/*func (s *Snapshot) NumSites() int {
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/internal/cache/snapshot.go:110]: Remove it

File: globalscheduler/pkg/scheduler/router/router.go [hunk 14]
```diff
@@ -17,13 +17,14 @@ limitations under the License.
 package router
 
 import (
-	"net"
+	//"net"
 	"strings"
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/router/router.go:21]: Remove it

File: globalscheduler/pkg/scheduler/router/router.go [hunk 15]
```diff
@@ -37,15 +38,16 @@ type Route struct {
...
-	apiVersion := "/" + config.DefaultString("component_version", "v1")
-
+	//apiVersion := "/" + config.DefaultString("component_version", "v1")
 	ws := new(restful.WebService)
-	ws.Path(apiVersion).Consumes(restful.MIME_JSON, restful.MIME_XML).Produces(restful.MIME_JSON, restful.MIME_XML)
+//	ws.Path(apiVersion).Consumes(restful.MIME_JSON, restful.MIME_XML).Produces(restful.MIME_JSON, restful.MIME_XML)
+	ws.Path("/").Consumes("*/*").Produces("*/*")
 	registerRoute(ws)
 	restful.Add(ws)
+//	klog.Infof("*** HTTP apiVersion ==> %v, path ==> %v", apiVersion, ws.Path(apiVersion))
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/router/router.go:47]: Remove it

File: globalscheduler/pkg/scheduler/router/router.go [hunk 16]
```diff
@@ -87,4 +89,18 @@ var routes = Routes{
...
 	},
+	Route{
+		"PushSnapshot",
+		strings.ToUpper("Patch"),
+		//http.MethodPatch,
+		//"/globalscheduler/v1/regionresources/regionname",
+		"/globalscheduler/v1/regionresources/{regionname}",
+		service.PushSnapshot,
+	},
+	Route{
```
Comment (jshaofuturewei) [inline globalscheduler/pkg/scheduler/router/router.go:100]: Remove it

File: globalscheduler/pkg/scheduler/common/apiserver/server.go (hunk 1 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/common/apiserver/server.go:50]: the port is needed

File: globalscheduler/pkg/scheduler/router/router.go (hunk 16 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/router/router.go:100]: Removed

File: globalscheduler/pkg/scheduler/router/router.go (hunk 15 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/router/router.go:47]: Removed

File: globalscheduler/pkg/scheduler/router/router.go (hunk 14 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/router/router.go:21]: Removed

File: globalscheduler/pkg/scheduler/internal/cache/snapshot.go (hunk 13 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/internal/cache/snapshot.go:110]: removed

File: globalscheduler/pkg/scheduler/internal/cache/snapshot.go (hunk 12 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/internal/cache/snapshot.go:57]: Removed

File: globalscheduler/pkg/scheduler/eventhandlers.go (hunk 5 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/eventhandlers.go:424]: removed V(3)

File: globalscheduler/pkg/scheduler/eventhandlers.go (hunk 4 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/eventhandlers.go:376]: removed V(3)

File: globalscheduler/pkg/scheduler/eventhandlers.go (hunk 3 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/eventhandlers.go:278]: Removed

File: globalscheduler/pkg/scheduler/common/apiserver/server.go (hunk 2 above)
Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/common/apiserver/server.go:143]: changed to Errorf

File: globalscheduler/pkg/scheduler/common/apiserver/server.go (hunk 1 above)
Comment (jshaofuturewei) [reply globalscheduler/pkg/scheduler/common/apiserver/server.go:49]: I mean we don't need "http(https)". We can use port instead. Sorry for the confusion

Comment (kimeunju108) [reply globalscheduler/pkg/scheduler/common/apiserver/server.go]: no problem.. the http is removed.

**Review by cathyhongzhang — APPROVED** [review]

**Review by jshaofuturewei — APPROVED** [review]

**Pull request merged**